backend/src/resolvers.js

```diff
@@ -1,127 +1,315 @@
+            connect: { id: args.post.id },
+          },
+        },
+      };
+
+      // check if user has voted and post is downvoted (-1)#
+      // if so , do nothing
+      if (votes.length > 0 && votes[0].value == -1) {
+        console.log('[downvote] user has downvoted, doing nothing...');
+
+        return;
+      }
+
+      // check if user has voted and post is upvoted (1)
+      // if so, update value on vote
+      const doesPostExistAndIsUpvoted = votes.length > 0 && votes[0].value == 1;
+
+      if (doesPostExistAndIsUpvoted) {
+        console.log('[downvote] post is upvoted, updating...');
 
-      if (!post) {
-        throw new UserInputError('Invalid post', { invalidArgs: args.postId });
+        // updat query
+        vote.where = { id: votes[0].id };
       }
 
-      return context.dataSources.db.downvotePost(post.id, context.req.auth.id);
+      // otherwise create a new vote
+      return await delegateToSchema({
+        schema,
+        operation: 'mutation',
+        fieldName: doesPostExistAndIsUpvoted ? 'updateVote' : 'createVote',
+        args: vote,
+        context,
+        info,
+      });
     },
 
+    delete: async (_parent, args, context, info) => {
+      const {
+        data: { posts },
+      } = await executor({
+        document: QUERY_DOES_AUTHOR_HAS_POSTS,
+        variables: {
+          postId: args.post.id,
+          userId: context.person.id,
+        },
+      });
 
-    delete: (parent, args, context) => {
-      return context.dataSources.db.deletePost(args.id);
+      // throw error if no posts avilable
+      if (!posts.length) {
+        throw new UserInputError(
+          'This is not your post or post does not exist!'
+        );
+      }
+
+      // delete all associated votes
+      const {
+        data: {
+          deleteManyVotesConnection: { pageInfo },
+        },
+      } = await executor({
+        document: DELETE_MANY_VOTES,
+        variables: { postId: args.post.id }
+      });
+
+      console.log('deletedCount', pageInfo.pageSize);
+
+      return await delegateToSchema({
+        schema,
+        operation: 'mutation',
+        fieldName: 'deletePost',
+        args: {
+          stage: 'DRAFT',
+          where: {
+            id: args.post.id,
+          },
+        },
+        context,
+        info,
+      });
     },
   },
 
-  Post: {
-    author(obj, args, context) {
-      return context.dataSources.db.getUser(obj.authorid);
-    },
-    votes(obj) {
-      return obj.voters.size;
+  Person: {
+    postCount: {
+      selectionSet: '{ posts { id } }',
+      resolve: (person) => person.posts.length,
     },
   },
 
-  //Users - allPosts
-  User: {
-    posts(obj, args, context) {
-      return context.dataSources.db
-        .allPosts()
-        .filter((post) => post.authorid === obj.id);
+  Post: {
+    votesCount: {
+      selectionSet: '{ votes {value} }',
+      resolve: (post) =>
+        post.votes.map((v) => v.value).reduce((a, b) => (a += b), 0),
```

Comment:
I would have modeled the schema with 2 relationships instead of 1 relationship with a relationship attribute. Then you could simply do `upvotersCount - downvotersCount` and the potential for invalid states is lower (e.g. you cannot set the value to `0` or values greater than `1`).